lib/internal/Magento/Framework/Controller/Result/Redirect.php

```diff
@@ -93,11 +100,27 @@ public function setPath($path, array $params = [])
     }
 
     /**
+     * Checks if current status code is a valid redirect code
+     * @param mixed $statusCode
+     * @return bool
+     */
+    protected function isValidHttpResponseCode($statusCode)
+    {
+        $statusCode = (int) $statusCode;
+        return !empty($statusCode) &&
+            in_array($statusCode, self::REDIRECT_RESPONSE_STATUS_CODES);
+    }
+
+    /**
      * {@inheritdoc}
      */
     protected function render(HttpResponseInterface $response)
     {
-        $response->setRedirect($this->url);
+        if ($this->isValidHttpResponseCode($this->httpResponseCode)) {
```

Comment:
Please replace to `if (!empty($this->httpResponseCode)) {`

as there is already similar check in ZF

```
if ((300 <= $code) && (307 >= $code)) {
            $this->_isRedirect = true;
        } else {
            $this->_isRedirect = false;
        }
```